pkg/service/utils.go

```diff
@@ -159,3 +161,55 @@ func (s *service) convertComponentRunToPB(run datamodel.ComponentRun) (*pipeline
 func CanViewPrivateData(namespace, requesterUID string) bool {
 	return namespace == requesterUID
 }
+
+func parseMetadataToStructArray(metadataMap map[string][]byte, log *zap.Logger, key string, metadataType string, logFields ...zap.Field) []*structpb.Struct {
+	md, ok := metadataMap[key]
+	if !ok {
+		log.Error(fmt.Sprintf("failed to load %s metadata", metadataType), logFields...)
+		return nil
+	}
+
+	structArr := make([]*structpb.Struct, 0)
+	if err := json.Unmarshal(md, &structArr); err != nil {
+		log.Error(fmt.Sprintf("failed to parse %s metadata", metadataType), logFields...)
+		return nil
+	}
+
+	return structArr
+}
+
+func parseRecipeMetadata(metadataMap map[string][]byte, log *zap.Logger, converter Converter, key string, metadataType string, logFields ...zap.Field) (*structpb.Struct, *pipelinepb.DataSpecification) {
+	md, ok := metadataMap[key]
+	if !ok {
+		log.Error(fmt.Sprintf("failed to load %s metadata", metadataType), logFields...)
```

Comment:
A few notes on log messages and fields:
- Go errors shouldn't be capitalised because they are meant to build an error stack: `parsing pipeline run metadata: key doesn't exist`. Log messages, however, won't be stacked with other logs, so a more human format is preferred, e.g., `Failed to load metadata`.
- Try to avoid building log messages with parameters. That's what fields are for. When you explore the logs, many times you want to group them by the log message. Usually the log explorer service (e.g. DataDog) will do this for you and display the occurrences and trends over a period of time. By parametrising the message you're ungrouping these logs. And, if you want to filter the results by the metadata type, leveraging a dedicated field will always be more efficient than a regexp on the log message.